GVFS/GVFS.Common/LibGit2RepoPool.cs

```diff
@@ -1,47 +1,66 @@
 ﻿using GVFS.Common.Git;
 using GVFS.Common.Tracing;
 using System;
-using System.Collections.Concurrent;
 using System.Threading;
 
 namespace GVFS.Common
 {
     public class LibGit2RepoPool
     {
-        private const int TryAddTimeoutMilliseconds = 10;
-        private const int TryTakeTimeoutMilliseconds = Timeout.Infinite;
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
 
-        private readonly BlockingCollection<LibGit2Repo> pool;
+        private readonly Func<LibGit2Repo> createRepo;
         private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private bool stopped;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+        private int activeCallers;
 
-        public LibGit2RepoPool(ITracer tracer, Func<LibGit2Repo> createRepo, int size)
+        public LibGit2RepoPool(ITracer tracer, Func<LibGit2Repo> createRepo, int size = 1, TimeSpan? disposalPeriod = null)
         {
             if (size <= 0)
             {
                 throw new ArgumentException("ProcessPool: size must be greater than 0");
             }
 
             this.tracer = tracer;
-            this.pool = new BlockingCollection<LibGit2Repo>();
-            for (int i = 0; i < size; ++i)
-            {
-                this.pool.Add(createRepo());
-            }
+            this.createRepo = createRepo;
+
+            this.sharedRepositoryDisposalPeriod = disposalPeriod ?? DefaultRepositoryDisposalPeriod;
+
+            this.sharedRepoDisposalTimer = new Timer(
+                (state) => this.DisposeSharedRepo(),
+                state: null,
+                dueTime: this.sharedRepositoryDisposalPeriod,
+                period: this.sharedRepositoryDisposalPeriod);
         }
 
         public void Dispose()
         {
-            this.pool.CompleteAdding();
-            this.CleanUpPool();
+            this.stopped = true;
+
+            lock (this.sharedRepoLock)
+            {
+                this.sharedRepo?.Dispose();
+                this.sharedRepo = null;
+            }
         }
 
         public bool TryInvoke<TResult>(Func<LibGit2Repo, TResult> function, out TResult result)
         {
-            LibGit2Repo repo = null;
+            if (this.stopped)

```

Comment:
Is reading/writing the value of `this.stopped` safe with multiple threads? Should we not be in a lock, or using `Interlocked`? (I know there is no `bool` overload for compare/exchange atomic operations, but we could just use an `int` for the Boolean in that case.)

Reply:
I'll make it `volatile` for extra help, but this check is actually useless. We check `this.stopped` under a lock in `GetRepoFromPool()`.